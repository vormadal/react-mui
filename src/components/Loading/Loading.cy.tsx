import React from 'react'
import Loading from './Loading'

describe('<Loading />', () => {
  it('Show spinner when loading', () => {
    // see: https://on.cypress.io/mounting-react
    cy.mount(<Loading loading />)
    cy.get('.MuiCircularProgress-root').should('exist')
  })
  it('Show data only when loading and containing data', () => {
    cy.mount(
      <Loading
        loading
        data={'content'}
      >
        {(data) => <div data-cy="container">{data}</div>}
      </Loading>
    )

    cy.get('[data-cy=container]').should('have.text', 'content')
    cy.get('.MuiCircularProgress-root').should('not.exist')
  })
  it('Show error message done loading', () => {
    cy.mount(<Loading error="this is an error">{() => <div data-cy="container">content</div>}</Loading>)

    cy.get('[data-cy=container]').should('not.exist')
    cy.get('.MuiCircularProgress-root').should('not.exist')
    cy.get('.MuiAlert-message').should('have.text', 'this is an error')
  })
})
