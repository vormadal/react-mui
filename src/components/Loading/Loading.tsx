import { Alert, Button } from '@mui/material'
import Spinner, { SpinnerSize } from '../Spinner'

type Props<T> = {
  children?: (data: T) => React.ReactElement
  loading?: boolean
  error?: string
  retry?: () => void | Promise<void>
  data?: T
  showReloads?: boolean
}

function Loading<T>({ children, loading, error, retry, data, showReloads }: Props<T>) {
  if (loading && (!data || showReloads)) return <Spinner show />
  if (error)
    return (
      <Alert
        severity="error"
        action={
          retry ? (
            <Button
              color="inherit"
              size="small"
              onClick={retry}
            >
              Retry
            </Button>
          ) : null
        }
      >
        {error}
      </Alert>
    )
  if (data && children)
    return (
      <>
        {children(data)}
        {loading && (
          <div style={{ position: 'absolute' }}>
            <Spinner
              show
              size={SpinnerSize.small}
            />
          </div>
        )}
      </>
    )
  return null
}

export default Loading
