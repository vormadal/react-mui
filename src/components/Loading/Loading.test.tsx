import * as renderer from 'react-test-renderer'
import Loading from './Loading'

describe('Loading component', () => {
  it('Show spinner when loading even if error is specified', () => {
    const component = renderer.create(
      <Loading
        loading
        error="error"
      >
        {() => <h1>test</h1>}
      </Loading>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Show small spinner when loading but already containing data', () => {
    const component = renderer.create(
      <Loading
        loading
        data="content"
      >
        {(data) => <h1>{data}</h1>}
      </Loading>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
