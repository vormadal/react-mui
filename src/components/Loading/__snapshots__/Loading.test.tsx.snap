// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component Show small spinner when loading but already containing data 1`] = `
[
  <h1>
    content
  </h1>,
  <div
    style={
      {
        "position": "absolute",
      }
    }
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-18lrjg1-MuiCircularProgress-root{display:inline-block;color:#1976d2;-webkit-animation:animation-61bdi0 1.4s linear infinite;animation:animation-61bdi0 1.4s linear infinite;}@-webkit-keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}@keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
        }
      }
      data-emotion="css 18lrjg1-MuiCircularProgress-root animation-61bdi0"
    />
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style={
        {
          "height": 20,
          "width": 20,
        }
      }
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-1idz92c-MuiCircularProgress-svg{display:block;}",
          }
        }
        data-emotion="css 1idz92c-MuiCircularProgress-svg"
      />
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <style
          dangerouslySetInnerHTML={
            {
              "__html": ".css-176wh8e-MuiCircularProgress-circle{stroke:currentColor;stroke-dasharray:80px,200px;stroke-dashoffset:0;-webkit-animation:animation-1p2h4ri 1.4s ease-in-out infinite;animation:animation-1p2h4ri 1.4s ease-in-out infinite;}@-webkit-keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}@keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}",
            }
          }
          data-emotion="css 176wh8e-MuiCircularProgress-circle animation-1p2h4ri"
        />
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={{}}
        />
      </svg>
    </span>
  </div>,
]
`;

exports[`Loading component Show spinner when loading even if error is specified 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-18lrjg1-MuiCircularProgress-root{display:inline-block;color:#1976d2;-webkit-animation:animation-61bdi0 1.4s linear infinite;animation:animation-61bdi0 1.4s linear infinite;}@-webkit-keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}@keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
      }
    }
    data-emotion="css 18lrjg1-MuiCircularProgress-root animation-61bdi0"
  />,
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-1idz92c-MuiCircularProgress-svg{display:block;}",
        }
      }
      data-emotion="css 1idz92c-MuiCircularProgress-svg"
    />
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-176wh8e-MuiCircularProgress-circle{stroke:currentColor;stroke-dasharray:80px,200px;stroke-dashoffset:0;-webkit-animation:animation-1p2h4ri 1.4s ease-in-out infinite;animation:animation-1p2h4ri 1.4s ease-in-out infinite;}@-webkit-keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}@keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}",
          }
        }
        data-emotion="css 176wh8e-MuiCircularProgress-circle animation-1p2h4ri"
      />
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={{}}
      />
    </svg>
  </span>,
]
`;

exports[`Show spinner when loading even if error is specified 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-18lrjg1-MuiCircularProgress-root{display:inline-block;color:#1976d2;-webkit-animation:animation-61bdi0 1.4s linear infinite;animation:animation-61bdi0 1.4s linear infinite;}@-webkit-keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}@keyframes animation-61bdi0{0%{-webkit-transform:rotate(0deg);-moz-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}100%{-webkit-transform:rotate(360deg);-moz-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
      }
    }
    data-emotion="css 18lrjg1-MuiCircularProgress-root animation-61bdi0"
  />,
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      {
        "height": 40,
        "width": 40,
      }
    }
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-1idz92c-MuiCircularProgress-svg{display:block;}",
        }
      }
      data-emotion="css 1idz92c-MuiCircularProgress-svg"
    />
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <style
        dangerouslySetInnerHTML={
          {
            "__html": ".css-176wh8e-MuiCircularProgress-circle{stroke:currentColor;stroke-dasharray:80px,200px;stroke-dashoffset:0;-webkit-animation:animation-1p2h4ri 1.4s ease-in-out infinite;animation:animation-1p2h4ri 1.4s ease-in-out infinite;}@-webkit-keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}@keyframes animation-1p2h4ri{0%{stroke-dasharray:1px,200px;stroke-dashoffset:0;}50%{stroke-dasharray:100px,200px;stroke-dashoffset:-15px;}100%{stroke-dasharray:100px,200px;stroke-dashoffset:-125px;}}",
          }
        }
        data-emotion="css 176wh8e-MuiCircularProgress-circle animation-1p2h4ri"
      />
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={{}}
      />
    </svg>
  </span>,
]
`;
